### PR TITLE
[Quant][fx] Remove WEIGHT_INDEX_DICT and BIAS_INDEX_DICT (Part 2)

### LINKED ISSUE #83093: Remove WEIGHT_INDEX_DICT and BIAS_INDEX_DICT from fx quant codebase
## Context
FX Graph Mode Quantization is an automated flow for quantization (https://pytorch.org/docs/stable/quantization.html#prototype-fx-graph-mode-quantization), since we may target different backends that supports quantization, and different backends have different support for quantized ops/dtypes etc. we developed `BackendConfig` that is used to configure quantization flow for various backends. More details in https://github.com/pytorch/rfcs/blob/master/RFC-0019-Extending-PyTorch-Quantization-to-Custom-Backends.md

The main goal for this task is to make some parts of hardcoded things configurable.

## Task
For a function call like: F.conv2d(input, weight, bias, ...), we need to identify which argument is weight, bias or input, right now we have two hardcoded dictionaries that's used to identify the weight and bias for different operators: WEIGHT_INDEX_DICT (https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fx/utils.py#L57) and BIAS_INDEX_DICT (https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fx/utils.py#L69). most of the other configurations are defined in https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/backend_config/native.py#L243

We should use BackendConfig to define these configurations as well, in this case, we can use https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/backend_config/backend_config.py#L305 (_set_input_type_to_index) to define the map from "weight"/"bias" to the index of the argument in the operator. e.g.
F.linear(input, weight, bias), then {"input": 0, "weight": 1, "bias": 2}, example: https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/backend_config/tensorrt.py#L33, we can move these configurations to the operator configurations in https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/backend_config/_common_operator_config_utils.py#L200,

e.g.
```
# this in a configuration for F.conv1d
conv_configs.append(
            BackendPatternConfig(convs.func)
                .set_observation_type(observation_type)  # noqa: E131
                .set_dtype_configs(dtype_configs))
                . _set_input_type_to_index({"weight": 1, "bias": 2})
# and do this for all the operators in WEIGHT_INDEX_DICT and BIAS_INDEX_DICT
```

in the end we should be able to remove the use of WEIGHT_INDEX_DICT and BIAS_INDEX_DICT in https://github.com/pytorch/pytorch/blob/master/torch/ao/quantization/fx/prepare.py, and derive these two dictionaries from configurations.

### Versions

master

cc @jerryzh168 @jianyuh @raghuramank100 @jamesr66a @vkuzo @ezyang @SherlockNoMad

### PR DESCRIPTION
**Pull request by jcaip**
Summary:
- Finishes the second part of https://github.com/pytorch/pytorch/pull/83263
- Removes WEIGHT_INDEX_DICT and BIAS_INDEX_DICT from utils.py
- Moves two funcitons, `node_arg_is_weight` and `node_arg_is_bias` into utils.py from prepare.py
convert.py and _equalize.py now use node_arg_is_weight instead of the dictionaries
- Adds in quantization support for `F.groupnorm`.

Add in missing BackendPatternConfigs for layernorm, instancenorm, and groupnorm

Test Plan:
```
python test/test_quantization.py TestQuantizeFx
python test/test_quantization.py TestQuantizeFxOps
```

Reviewers:

Subscribers:

Tasks:

Tags:

ghstack-source-id: 2b157e0dc4f1553be1f4813b4693db952e6fc558
Pull Request resolved: https://github.com/pytorch/pytorch/pull/83848

Fixes #83093

### CODE REVIEW

File: torch/ao/ns/fx/mappings.py [hunk 1]
```diff
@@ -195,6 +195,10 @@ def get_base_name_to_sets_of_related_ops() -> Dict[str, Set[NSNodeTargetType]]:
         set([
             F.hardswish,
         ]),
+        # F.group_norm
```
Comment (jcaip) [inline torch/ao/ns/fx/mappings.py:198]: I added this to get a test to pass, but I don't know if I need to do anything else other than modify the dict

**Comment by andrewor14** [comment]
Can you add (part 2) to the title to differentiate this from the previous PR?

File: torch/ao/quantization/fx/utils.py [hunk 2]
```diff
@@ -38,43 +39,40 @@
...
-    torch.nn.functional.conv1d : [2],
-    torch.nn.functional.conv2d : [2],
-    torch.nn.functional.conv3d : [2],
-    torch.nn.functional.linear : [2],
-    torch.nn.functional.layer_norm : [3],
-    torch.nn.functional.group_norm : [3],
-    torch.nn.functional.instance_norm : [4],
-}
+def node_arg_is_weight(node: Node, arg: Any, backend_config: BackendConfig) -> bool:
+    "Returns if node arg is weight"
```
Comment (andrewor14) [inline torch/ao/quantization/fx/utils.py:60]: nit: use 3 double quotes here:
```
"""Returns if ... """
```

File: torch/ao/quantization/fx/_equalize.py [hunk 3]
```diff
@@ -305,9 +307,10 @@ def get_op_node_and_weight_eq_obs(
 def maybe_get_weight_eq_obs_node(op_node: Node, modules: Dict[str, nn.Module]) -> Optional[Node]:
     """ Gets the weight equalization observer node if it exists.
     """
-    assert(op_node.op == 'call_function' and op_node.target in WEIGHT_INDEX_DICT)
-    for i, node_arg in enumerate(op_node.args):
-        if i in WEIGHT_INDEX_DICT[op_node.target]:  # type: ignore[index]
+    assert(op_node.op == 'call_function')
+    backend_config = get_native_backend_config()
```
Comment (andrewor14) [inline torch/ao/quantization/fx/_equalize.py:311]: I think we should add a TODO here to pass in the backend_config here instead

**Review by andrewor14 — COMMENTED** [review]
Looks great! Just left some minor comments

File: torch/ao/ns/fx/mappings.py (hunk 1 above)
Comment (jerryzh168) [reply torch/ao/ns/fx/mappings.py:198]: not sure why we need this, cc @vkuzo to check

File: torch/ao/quantization/backend_config/_common_operator_config_utils.py [hunk 4]
```diff
@@ -350,16 +350,35 @@ def _get_default_op_configs(dtype_configs: List[DTypeConfig]) -> List[BackendPat
...
+
+    configs.append(
+        BackendPatternConfig(torch.nn.functional.layer_norm)
+        .set_observation_type(ObservationType.OUTPUT_USE_DIFFERENT_OBSERVER_AS_INPUT)  # noqa: E131
+        .set_dtype_configs(dtype_configs)
+        ._set_input_type_to_index({"weight": 2, "bias": 3})
+    )
+
+    configs.append(
+        BackendPatternConfig(torch.nn.functional.group_norm)
```
Comment (jerryzh168) [inline torch/ao/quantization/backend_config/_common_operator_config_utils.py:370]: oh, I think we added quantization support for group_norm here, group_norm was not supported in the BackendConfig previously (since we removed L353 and L356 only, there were no entry for group_norm there)

**Review by jerryzh168 — APPROVED** [review]
looks great, thanks for removing the hardcode!

File: torch/ao/quantization/backend_config/_common_operator_config_utils.py (hunk 4 above)
Comment (vkuzo) [reply torch/ao/quantization/backend_config/_common_operator_config_utils.py:375]: Quantized group_norm, instance_norm{n}d, layer_norm is supported in fbgemm but not qnnpack/xnnpack.  It would be great to move these config entries to the fbgemm section.

Comment (jcaip) [reply torch/ao/quantization/backend_config/_common_operator_config_utils.py:375]: @vkuzo Should I pull out the `nn.Modules` too for instanceNorm/LayerNorm or just the functionals?

Comment (vkuzo) [reply torch/ao/quantization/backend_config/_common_operator_config_utils.py]: Sorry, I made an outdated and incorrect statement.  What I should have said is that group_norm, instance_norm{n}d, layer_norm have an optimized quantized implementation on fbgemm, and have a naive implementation on xnnpack/qnnpack.  Please ignore my comment above.

File: torch/ao/quantization/backend_config/fbgemm.py [hunk 5]
```diff
@@ -71,6 +72,49 @@
...
+
+    configs.append(
+        BackendPatternConfig(torch.nn.functional.group_norm)
+        .set_observation_type(ObservationType.OUTPUT_USE_DIFFERENT_OBSERVER_AS_INPUT)  # noqa: E131
+        .set_dtype_configs(dtype_configs)
+        ._set_input_type_to_index({"weight": 2, "bias": 3})
+    )
+
+    configs.append(
+        BackendPatternConfig(torch.nn.functional.instance_norm)
```
Comment (andrewor14) [inline torch/ao/quantization/backend_config/fbgemm.py:111]: Per @vkuzo's comment [here](https://github.com/pytorch/pytorch/pull/83853#discussion_r956261560), should we move these back to `_common_operator_config_utils.py`?

File: torch/ao/quantization/backend_config/fbgemm.py [hunk 6]
```diff
@@ -71,6 +72,49 @@
...
+
+    configs.append(
+        BackendPatternConfig(torch.nn.functional.layer_norm)
+        .set_observation_type(ObservationType.OUTPUT_USE_DIFFERENT_OBSERVER_AS_INPUT)  # noqa: E131
+        .set_dtype_configs(dtype_configs)
+        ._set_input_type_to_index({"weight": 2, "bias": 3})
+    )
+
+    configs.append(
+        BackendPatternConfig(torch.nn.functional.group_norm)
```
Comment (andrewor14) [inline torch/ao/quantization/backend_config/fbgemm.py:104]: I guess this changes functionality, since `group_norm` was previously not defined in `BackendConfig`? Maybe we should mention this in the PR description to be clear.

File: torch/ao/quantization/backend_config/fbgemm.py [hunk 7]
```diff
@@ -71,6 +72,49 @@
...
+# ===================
+
+
+def _get_fbgemm_op_configs(dtype_configs: List[DTypeConfig]) -> List[BackendPatternConfig]:
+    configs = []
+    fbgemm_ops = [
+        torch.nn.InstanceNorm1d,
+        torch.nn.InstanceNorm2d,
+        torch.nn.InstanceNorm3d,
+        torch.nn.LayerNorm,
```
Comment (andrewor14) [inline torch/ao/quantization/backend_config/fbgemm.py:86]: If we're adding functional `group_norm` anyway, should we also add `torch.nn.GroupNorm` here? @jerryzh168

**Review by andrewor14 — APPROVED** [review]
Looks great. Thanks!

File: torch/ao/quantization/backend_config/fbgemm.py (hunk 7 above)
Comment (jerryzh168) [reply torch/ao/quantization/backend_config/fbgemm.py]: yeah I think we can do that since we have quantized module for these things: https://github.com/pytorch/pytorch/blob/master/torch/ao/nn/quantized/modules/normalization.py, maybe better to do this in a separate PR.

Comment (jcaip) [reply torch/ao/quantization/backend_config/fbgemm.py]: I can make a PR to add it in later.
